### DISCUSSION TITLE
What is a 'A3DNewtonianLiquid.dll".

### DISCUSSION THREAD

**Discussion by Antikira** [Anura3D/Anura3D_OpenSource#54] [Q&A] [answered]
I want to investigate the scouring phenomenon caused by water.
At this time, the 'A3DNewtonianLiquid.dll' file appears below when setting the water properties, but I could not find this file.
What kind of file is this 'A3DNewtonianLiquid.dll' and is it necessary to calculate the appearance of water?

**Comment by zhangfengeo** [comment]
it seems that the 'A3DNewtonianLiquid.dll' file isn't involved in the source code of Anura3D,maybe you can use 'A3DBingham.dll' as an substitution,which is a kind of non-newtonian fluid model.

**Comment by murphy-john** [comment, marked as answer]
@Antikira to answer your questions:
1) A "*.dll" file is a Dynamic Link Library file. You can search online if you'd like more information. In Anura3D, we used DLL files for external material models (i.e. solid (e.g. soil) models, liquid models, etc.). Using a DLL file allows users to create an external material model to calculate stresses in an external procedure without having to make edits to the main code. In this case, the note with "A3DNewtonianLiquid.dll" you are mentioning is in the GID interface as a placeholder. 

Some additional background: When the code was originally developed, all the material models were coded directly into the main source code. When the Anura3D developers community added the ability to use external material models in dlls, some of the most common material models like, the "newtonian_liquid", remained in the source code because we haven't had time to transfer them into a DLL. This is because it takes a fair amount of time to move the code into a DLL and test it. We hope to have "newtonian_fluid" material model fully moved into the "A3DNewtonian_Liquid.DLL" in a future release. That leads to your next question...see below.

2) No, the "A3DNewtonian_Liquid.DLL" file is not required to run a model in the current open source release of the Anura3D code. You can specify the "Material model liquid" as "Newtonian liquid" and Anura3D will calculate the stresses using the main Anura3D code. You can find places where the liquid stresses are calculated in the "ModMPMDYNStresses" and "ModLiquid". It's up to the user to make sure the stresses are being calculated correctly for their model.

**Reply by qwerty12345697** [reply]
How can we specify the "Material model liquid" as "Newtonian liquid" in Anura3D, i don't know how to aviod the error.

**Reply by ayerrocolom** [reply]
If you want to simulate the material behavior based on a constitutive model different to those provided in Anura3D (i.e. linear elasticity, Mohr-Coulomb, Bingham),  you need to code your own material model following the instructions provided in the tutorial manual in chapter 4.
